homeassistant/components/discord/calendar.py

```diff
@@ -0,0 +1,101 @@
+"""Module for using the calendar entity platform."""
+
+from __future__ import annotations
+
+import datetime
+from typing import Any
+
+import nextcord
+
+from homeassistant.components.calendar import CalendarEntity, CalendarEvent
+from homeassistant.config_entries import ConfigEntry
+from homeassistant.core import HomeAssistant
+from homeassistant.helpers.entity_platform import AddEntitiesCallback
+
+from .const import DOMAIN
+
+PARALLEL_UPDATES = 1
+SCAN_INTERVAL = datetime.timedelta(minutes=10)
+
+
+async def async_setup_entry(
+    hass: HomeAssistant,
+    entry: ConfigEntry,
+    async_add_entities: AddEntitiesCallback,
+) -> None:
+    """Initialize the calendar entities from the configuration guilds."""
+    client = hass.data[DOMAIN][entry.entry_id]["client"]
+    async_add_entities(
+        [DiscordCalendar(guild) async for guild in client.fetch_guilds()],
+        False,
+    )
+
+
+class DiscordCalendar(CalendarEntity):
+    """A :py:class:`CalendarEntity` subclass that grabs events from discord."""
+
+    def __init__(self, guild: nextcord.Guild) -> None:
+        """Initialize attributes for :py:class:`DiscordCalendar`."""
+        self._guild = guild
+        self._events: list[CalendarEvent] = []
+        self.entity_id = f"calendar.{DOMAIN}_{guild.id}"
+
+    @property
+    def unique_id(self) -> str:
+        """Uniquely identify each calendar with the ID of its corresponding guild."""
+        return str(self._guild.id)
+
+    @property
+    def name(self) -> str:
+        """Display the name of the corresponding guild."""
+        return self._guild.name
+
+    @property
+    def event(self) -> CalendarEvent | None:
+        """Return the oldest event fetched most recently."""
+        if self._events:
+            return self._events[0]
+        return None
+
+    @property
+    def entity_picture(self) -> str:
+        """Return the Bot's profile picture to use in the frontend."""
+        if self._guild.icon is None:
+            return ""
+        return self._guild.icon.url
+
+    @property
+    def extra_state_attributes(self) -> dict[str, Any] | None:
+        """Return extra information about the corresponding guild."""
+        return {"guild.id": str(self._guild.id)}
+
+    async def async_get_events(
+        self,
+        hass: HomeAssistant,
+        start_date: datetime.datetime,
+        end_date: datetime.datetime,
+    ) -> list[CalendarEvent]:
+        """Return all events known locally between two timestamps."""
+        return [
+            event
+            for event in self._events
+            if start_date < event.start < end_date  # Start and end are exclusive.
```

Comment:
If you have a calendar event from Noon to 1PM and we're querying from 12:30PM to 1:30PM this event will not be returned so an end event trigger would not work.

Reply:
One way that google [calendar api describes](https://developers.google.com/calendar/api/v3/reference/events/list) how to implement this is:
- start time (or time min) "Lower bound (exclusive) for an event's end time to filter by"
- end time (or time max) "Upper bound (exclusive) for an event's start time to filter by"

which i found helpful.

Reply:
I've decided to check if the `event.start` is in the time range OR if `time.end` is in the range, instead of AND.

Reply:
The only caveat with that is: If you have a calendar event from Noon to 1PM and we're querying from 12:30PM to 12:45PM it won't be returned as active.  Neither start date or end date is in the range, but the event is still active during that time.

Reply:
Hmm so we simplify our logic to `if query_start_time < event.end and event.start < query_end_time`